### ISSUE TITLE
SQLFluff isn't detecting rule L014 unquoted identifiers violation when the identifier contains a Jinja variable. Using Snowflake Dialect.

### ISSUE THREAD

**Issue by garydeng01** [sqlfluff/sqlfluff#5616]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When rule L014 for unquoted identifiers is configured with "extended_capitalisation_policy = lower" (or "extended_capitalisation_policy = upper"), sqlfluff won't detect violating unquoted identifiers if they contain a Jinja variable.

E.g. when policy is lower - **{{ env_type }}_MY_DB**.schema.my_table

### Expected Behaviour

SQLFluff flags/applies rule L014 unquoted identifiers violation for unquoted identifiers that contain a Jinja variable.

### Observed Behaviour

When linting such a file, sqlfluff output doesn't contain any reported violations:

![image](https://github.com/sqlfluff/sqlfluff/assets/49459007/4aa3e6e7-a705-45c5-b987-203727457284)

### How to reproduce

1. Have the following rules in the .sqlfluff file
```
[sqlfluff]
templater = jinja
dialect = snowflake

[sqlfluff:templater:jinja:context]
env_type = DEV

[sqlfluff:rules:capitalisation.identifiers] # L014 Unquoted Identifiers
extended_capitalisation_policy = lower
```
2. Have the following sql in example.sql
```
----------------------------
-- Set Jinja variables
----------------------------
{% set db_name = 'DB' %}
{% set schema_name = 'SCHEMA' %}
{% set role_name = 'ADMIN' %}

----------------------------
-- Set context
----------------------------
USE ROLE {{ env_type }}_{{ role_name }}_rl;
USE DATABASE {{ env_type }}_{{ db_name }};
USE SCHEMA {{ schema_name }};

----------------------------
-- Script
----------------------------
CREATE OR REPLACE VIEW my_view (
) AS
SELECT *
FROM {{ env_type }}_MY_DB.schema.my_table
WHERE column = 'Value';

3. Put .sqlfluff and example.sql in the same folder
4. Navigate your terminal into the folder containing those two files
5. Run `sqlfluff lint .\example.sql`

```

### Dialect

Snowflake

### Version

2.3.5

### Configuration

[sqlfluff]
templater = jinja
dialect = snowflake

[sqlfluff:templater:jinja:context]
env_type = DEV

[sqlfluff:rules:capitalisation.identifiers] # L014 Unquoted Identifiers
extended_capitalisation_policy = lower

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
